pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2154,13 +2180,39 @@ public boolean addNewReloadAllSegmentsJob(String tableNameWithType, String jobId
     jobMetadata.put(CommonConstants.ControllerJob.JOB_TYPE, ControllerJobType.RELOAD_ALL_SEGMENTS.toString());
     jobMetadata.put(CommonConstants.ControllerJob.SUBMISSION_TIME_MS, Long.toString(System.currentTimeMillis()));
     jobMetadata.put(CommonConstants.ControllerJob.MESSAGE_COUNT, Integer.toString(numberOfMessagesSent));
-    return addControllerJobToZK(jobId, jobMetadata);
+    return addControllerJobToZK(jobId, jobMetadata, ZKMetadataProvider.constructPropertyStorePathForControllerJob(),
+            _propertyStore);
   }
 
-  private boolean addControllerJobToZK(String jobId, Map<String, String> jobMetadata) {
-    String jobResourcePath = ZKMetadataProvider.constructPropertyStorePathForControllerJob();
+  public static boolean addRebalanceResultToZk(String tableNameWithType,
+                                               String jobId,
+                                               TableRebalanceProgressStats tableRebalanceProgressStats,
+                                               ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    Map<String, String> jobMetadata = new HashMap<>();
+    jobMetadata.put(CommonConstants.ControllerJob.TABLE_NAME_WITH_TYPE, tableNameWithType);
+    jobMetadata.put(CommonConstants.ControllerJob.JOB_ID, jobId);
+    jobMetadata.put(CommonConstants.ControllerJob.SUBMISSION_TIME_MS, Long.toString(System.currentTimeMillis()));
+    jobMetadata.put(CommonConstants.ControllerJob.JOB_TYPE, ControllerJobType.TABLE_REBALANCE.name());
+    try {
+      jobMetadata.put(RebalanceConfigConstants.REBALANCE_PROGRESS_STATS,
+              JsonUtils.objectToString(tableRebalanceProgressStats));
+    } catch (JsonProcessingException e) {
+      LOGGER.error("Error serialising rebalance stats to JSON for persisting to ZK {}",
+              jobId, e);
+    }
+    return addControllerJobToZK(jobId, jobMetadata,
+                                ZKMetadataProvider.constructPropertyStorePathForControllerJobType(
+                                ControllerJobType.TABLE_REBALANCE.name()),
+                                propertyStore);
+  }
+
+  private static boolean addControllerJobToZK(String jobId, Map<String, String> jobMetadata,
```

Comment:
Suggest not making it a static method. We may pass the `PinotHelixResourceManager` when constructing the `ZkBasedTableRebalanceObserver`. We should only construct `ZkBasedTableRebalanceObserver` from the `PinotHelixResourceManager` to ensure it is triggered by the controller.